### DISCUSSION TITLE
Csla 5 + DI + Testing

### DISCUSSION THREAD

**Discussion by RocketLabsDev** [MarimerLLC/csla#2779] [Questions] [answered]
Hello,

Just wondering how to set up unit tests with dependency injection.
Some examples would be great, for example in the faststart or projecttracker project.

Another thing, when using the business objects in an WPF project, where should I set up the services.

**Comment by TheCakeMonster** [comment]
Have a look at the Csla.Tests project in the main branch of the repo for examples of how to do DI for tests. We don't really need samples when we have a test suite that uses the technique for real.

You need DI less when using Csla 5, but that's not to say you don't need it. The principles are the same for both versions.

WPF isn't my thing, but configuration will follow the conventions for whatever host you are running. It's a WPF question rather than a CSLA question, unless I'm missing the point.

**Reply by TheCakeMonster** [reply]
Here's one example, of many that are present throughout the test project.

https://github.com/MarimerLLC/csla/blob/main/Source/Csla.test/Basic/BasicTests.cs

Note the use of the TestDIContext class, which I have created in an assembly called Csla.TestHelpers. This is created from a static class called TestDIContextFactory, which initialises a new DI container configured for use by CSLA each time it is used.

We're discussing whether maybe unit testing should be a first class target for CSLA, by creating a Csla.UnitTesting package. That discussion is [here](https://github.com/MarimerLLC/csla/discussions/2704) in case you want to join in.

**Comment by RocketLabsDev** [comment, marked as answer]
But if i have this kind of fetc how do i make a test for it when the resourcedal needs to be injected?
```    [Fetch]
    private void Fetch(int id, [Inject] IResourceDal dal)
    {
        var data = dal.Fetch(id);
      using (BypassPropertyChecks)
      {
        Id = data.Id;
        FirstName = data.FirstName;
        LastName = data.LastName;
        TimeStamp = data.LastChanged;
        Assignments = DataPortal.FetchChild<ResourceAssignments>(id);
      }
    }
```
The test method something like this:
```
    [TestMethod]
    public void ResourceTestFetch()
    {
      IResourceDal dal = new ProjectTracker.DalMock.ResourceDal();
      var result = Csla.DataPortal.Fetch<ResourceEdit>(OriginalId, dal);
      Assert.AreEqual(OriginalId, result.id);
    }
```
The dal needs to be injected so the test method obviously doesn't work.

**Reply by TheCakeMonster** [reply]
Yes, you need to register your data portal into the DI container which CSLA is using.

Here's an example of a test that registers a type into the DI container:

https://github.com/MarimerLLC/csla/blob/main/Source/Csla.test/DataPortal/AuthorizeDataPortalTests.cs

See line 43, where I register a type named TestableDataPortal. You need to register your DAL implementation in the same way, although using AddScoped might be more appropriate for you than AddTransient.

In my case I have chosen to share the TestDIContext across multiple tests, so the registration happens in an Initialize method rather than the test method itself. However, this is your choice; you can create a new TestDIContext as part of each individual test method if you prefer.

**Reply by RocketLabsDev** [reply]
Ok, so this looks promising.
Do i need to get this from github or is it in NUGET package?

**Reply by TheCakeMonster** [reply]
There's no Nuget package yet, no. We're still working on that.

The code is open sourced under the very permissive MIT license, so you are free to use it as you see fit. Take a copy for the time being if you'd like to use it as is.

**Reply by RocketLabsDev** [reply]
Thank you!
Got the idea with this and got it working.
